src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubDirectDStream.java

```diff
@@ -0,0 +1,268 @@
+                                     .setSubscription(ProjectSubscriptionName.of(projectId, subscriptionId).toString())
+                                     .build());
+      return snapshotName;
+    } catch (NotFoundException e) {
+      //Saved snapshot is not found. Create one and save it.
+      throw new RuntimeException(String.format(
+        "Saved snapshot %s not found. Please clear the application state to proceed. " +
+          "REST api for state deletion is namespaces/{namespace-id}/apps/{app-id}/state .",
+        snapshotName), e);
+    }
+  }
+
+  @Nullable
+  private String fetchSnapShot(String subscriptionId, io.cdap.cdap.etl.api.streaming.StreamingContext context) {
+    Optional<byte[]> state = null;
+    try {
+      state = context.getState(subscriptionId);
+    } catch (IOException e) {
+      throw new RuntimeException(String.format("Error fetching saved state for subscription %s.", subscriptionId), e);
+    }
+
+    if (!state.isPresent()) {
+      LOG.debug("No saved state for {}.", subscriptionId);
+      return null;
+    }
+
+    try {
+      Snapshot snapshot = Snapshot.parseFrom(state.get());
+      LOG.debug("Found existing snapshot {} .", snapshot.getName());
+      return snapshot.getName();
+    } catch (InvalidProtocolBufferException e) {
+      throw new RuntimeException(String.format("Error parsing saved state for subscription %s.", subscriptionId), e);
+    }
+  }
+
+  private String createAndSave(String projectId, String subscriptionId, String snapshotName,
+                               io.cdap.cdap.etl.api.streaming.StreamingContext context) {
+    LOG.debug("Creating snapshot {} in Pub/Sub .", snapshotName);
+    Snapshot snapshot = createSnapshot(projectId, subscriptionId, snapshotName);
+    saveSnapshot(snapshot, subscriptionId, context);
+    return snapshot.getName();
+  }
+
+  private Snapshot createSnapshot(String projectId, String subscriptionId, String snapshotName) {
+    CreateSnapshotRequest request = CreateSnapshotRequest.newBuilder()
+      .setName(ProjectSnapshotName.of(projectId, snapshotName).toString())
+      .setSubscription(ProjectSubscriptionName.of(projectId, subscriptionId).toString()).putAllLabels(new HashMap<>())
+      .build();
+    return subscriptionAdminClient.createSnapshot(request);
+  }
+
+  private void deleteSnapshot(String snapshotName) {
+    ProjectSnapshotName projectSnapshotName = ProjectSnapshotName.parse(snapshotName);
+    subscriptionAdminClient.deleteSnapshot(projectSnapshotName);
+  }
+
+  private void saveSnapshot(Snapshot snapshot, String subscription,
+                            io.cdap.cdap.etl.api.streaming.StreamingContext context) {
+    try {
+      LOG.debug("Saving snapshot {} in state .", snapshot.getName());
+      context.saveState(subscription, snapshot.toByteArray());
+    } catch (IOException e) {
+      throw new RuntimeException("Error saving state. ", e);
```

Comment:
think it would be better to have this method throws IOException instead of wrapping as a RuntimeException. Will make it easier to reason about what happens if this fails.

Reply:
Done.